packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -0,0 +1,125 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {Result} from '../../api/search/search/result';
+import {isArray, removeDuplicates} from '../../utils/utils';
+import {executeSearch, fetchMoreResults} from '../search/search-actions';
+import {registerFolding} from './folding-actions';
+import {
+  FoldedResult,
+  FoldingFields,
+  getFoldingInitialState,
+} from './folding-state';
+
+interface ResultWithFolding extends Result {
+  parentResult?: ResultWithFolding;
+  childResults?: ResultWithFolding[];
+}
+
+function getRawValue(result: ResultWithFolding, field: string) {
+  const rawValue = result.raw[field] as string | [string] | undefined;
+  if (isArray(rawValue)) {
+    return rawValue[0];
+  }
+  return rawValue;
+}
+
+function areDefinedAndEqual<T>(
+  value1: T | undefined,
+  value2: T | undefined
+): boolean {
+  return (value1 || value2) !== undefined && value1 === value2;
+}
+
+function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
+  const results = [result];
+  if (result.parentResult) {
+    results.unshift(result.parentResult);
+  }
+  if (result.childResults?.length) {
+    results.push(
+      ...result.childResults.flatMap((childResult) =>
+        flattenResultTree(childResult)
+      )
+    );
+  }
+  return results;
+}
+
+function foldResult(
+  source: ResultWithFolding,
+  results: ResultWithFolding[],
+  fields: FoldingFields
+): FoldedResult {
+  const sourceChildValue = getRawValue(source, fields.child);
+  return {
+    ...source,
+    children: sourceChildValue
+      ? results
+          .filter(
+            (result) =>
+              getRawValue(result, fields.child) !==
+                getRawValue(source, fields.child) &&
+              getRawValue(result, fields.parent) === sourceChildValue
+          )
+          .map((result) => foldResult(result, results, fields))
+      : [],
+  };
+}
+
+function createCollection(
+  relevantResult: ResultWithFolding,
+  fields: FoldingFields
+): FoldedResult {
+  const flattenedResults = removeDuplicates(
+    flattenResultTree(relevantResult),
+    (result) => result.uniqueId
+  );
+
+  const topLevelResult =
+    flattenedResults.find(
+      (result) =>
+        getRawValue(result, fields.parent) === undefined ||
+        areDefinedAndEqual(
+          getRawValue(result, fields.parent),
+          getRawValue(result, fields.child)
+        )
```

Comment:
In the doc, a child referencing itself as its own parent is the root, while in the search UI samples it's a child with no parent.

Reply:
It would be good to store the two halves of the condition in well-named variables to better communicate these two cases.